corporate/lib/stripe.py

```diff
@@ -1658,6 +1671,59 @@ def process_downgrade(self, plan: CustomerPlan) -> None:
         plan.status = CustomerPlan.ENDED
         plan.save(update_fields=["status"])
 
+    @override
+    def get_initial_upgrade_context(
+        self, initial_upgrade_request: InitialUpgradeRequest
+    ) -> Tuple[Optional[str], Optional[Dict[str, Any]]]:
+        customer = self.get_customer()
+        if self.is_sponsored_or_pending(customer):
+            return reverse("sponsorship_request"), None
+
+        billing_page_url = reverse("self_hosting_auth_redirect")
+        if customer is not None and (get_current_plan_by_customer(customer) is not None):
+            return billing_page_url, None
```

Comment:
I think the one thing that won't immediately work for using the shared copy is these redirects:

```
        if self.is_sponsored_or_pending(customer):                                                                      
            return reverse("sponsorship_request"), None                                                                 
                                                                                                                        
        billing_page_url = reverse("billing_home")                                                                      
```

They need to be configurable based on the `BillingSession` subclass. I think it might be the case that we can just construct these with constructions of the form `f"{self.remote_billing_url}/sponsorship"` rather than using `reverse`.

Reply:
Done.